File: test/org/stringtemplate/test/TestDictionaries.java
package org.stringtemplate.test;

import org.junit.Test;
import org.stringtemplate.STGroup;
import org.stringtemplate.ST;
import org.stringtemplate.Misc;
import static org.junit.Assert.assertEquals;

import java.util.Map;
import java.util.HashMap;

public class TestDictionaries extends BaseTest {
    @Test public void testDict() throws Exception {
        String templates =
                "group test;" +newline+
                "typeInit ::= [\"int\":\"0\", \"float\":\"0.0\"] "+newline+
                "var(type,name) ::= \"<type> <name> = <typeInit.(type)>;\""+newline
                ;
        Misc.writeFile(tmpdir, "test.stg", templates);
        STGroup group = STGroup.load(tmpdir+"/"+"test.stg");
        ST st = group.getInstanceOf("var");
        st.add("type", "int");
        st.add("name", "x");
        String expecting = "int x = 0;";
        String result = st.render();
        assertEquals(expecting, result);
    }

    @Test public void testDictValuesAreTemplates() throws Exception {
        String templates =
                "group test;" +newline+
                "typeInit ::= [\"int\":{0<w>}, \"float\":{0.0<w>}] "+newline+
                "var(type,w,name) ::= \"<type> <name> = <typeInit.(type)>;\""+newline
                ;
        Misc.writeFile(tmpdir, "test.stg", templates);
        STGroup group = STGroup.load(tmpdir+"/"+"test.stg");
        ST st = group.getInstanceOf("var");
        st.add("w", "L");
        st.add("type", "int");
        st.add("name", "x");
        String expecting = "int x = 0L;";
        String result = st.render();
        assertEquals(expecting, result);
    }

    @Test public void testDictKeyLookupViaTemplate() throws Exception {
        // Make sure we try rendering stuff to string if not found as regular object
        String templates =
                "group test;" +newline+
                "typeInit ::= [\"int\":{0<w>}, \"float\":{0.0<w>}] "+newline+
                "var(type,w,name) ::= \"<type> <name> = <typeInit.(type)>;\""+newline
                ;
        Misc.writeFile(tmpdir, "test.stg", templates);
        STGroup group = STGroup.load(tmpdir+"/"+"test.stg");
        ST st = group.getInstanceOf("var");
        st.add("w", "L");
        st.add("type", new ST("int"));
        st.add("name", "x");
        String expecting = "int x = 0L;";
        String result = st.render();
        assertEquals(expecting, result);
    }

    @Test public void testDictKeyLookupAsNonToStringableObject() throws Exception {
        // Make sure we try rendering stuff to string if not found as regular object
        String templates =
                "group test;" +newline+
                "foo(m,k) ::= \"<m.(k)>\""+newline
                ;
        Misc.writeFile(tmpdir, "test.stg", templates);
        STGroup group = STGroup.load(tmpdir+"/"+"test.stg");
        ST st = group.getInstanceOf("foo");
        Map<HashableUser,String> m = new HashMap<HashableUser,String>();
        m.put(new HashableUser(99,"parrt"), "first");
        m.put(new HashableUser(172036,"tombu"), "second");
        m.put(new HashableUser(391,"sriram"), "third");
        st.add("m", m);
        st.add("k", new HashableUser(172036,"tombu"));
        String expecting = "second";
        String result = st.render();
        assertEquals(expecting, result);
    }

    @Test public void testDictMissingDefaultValueIsEmpty() throws Exception {
        String templates =
                "group test;" +newline+
                "typeInit ::= [\"int\":\"0\", \"float\":\"0.0\"] "+newline+
                "var(type,w,name) ::= \"<type> <name> = <typeInit.(type)>;\""+newline
                ;
        Misc.writeFile(tmpdir, "test.stg", templates);
        STGroup group = STGroup.load(tmpdir+"/"+"test.stg");
        ST st = group.getInstanceOf("var");
        st.add("w", "L");
        st.add("type", "double"); // double not in typeInit map
        st.add("name", "x");
        String expecting = "double x = ;"; // weird, but tests default value is key
        String result = st.render();
        assertEquals(expecting, result);
    }

    @Test public void testDictHiddenByFormalArg() throws Exception {
        String templates =
                "group test;" +newline+
                "typeInit ::= [\"int\":\"0\", \"float\":\"0.0\"] "+newline+
                "var(typeInit,type,name) ::= \"<type> <name> = <typeInit.(type)>;\""+newline
                ;
        Misc.writeFile(tmpdir, "test.stg", templates);
        STGroup group = STGroup.load(tmpdir+"/"+"test.stg");
        ST st = group.getInstanceOf("var");
        st.add("type", "int");
        st.add("name", "x");
        String expecting = "int x = ;";
        String result = st.render();
        assertEquals(expecting, result);
    }

    @Test public void testDictEmptyValueAndAngleBracketStrings() throws Exception {
        String templates =
                "group test;" +newline+
                "typeInit ::= [\"int\":\"0\", \"float\":, \"double\":<<0.0L>>] "+newline+
                "var(type,name) ::= \"<type> <name> = <typeInit.(type)>;\""+newline
                ;
        Misc.writeFile(tmpdir, "test.stg", templates);
        STGroup group = STGroup.load(tmpdir+"/"+"test.stg");
        ST st = group.getInstanceOf("var");
        st.add("type", "float");
        st.add("name", "x");
        String expecting = "float x = ;";
        String result = st.render();
        assertEquals(expecting, result);
    }

    @Test public void testDictDefaultValue() throws Exception {
        String templates =
                "group test;" +newline+
                "typeInit ::= [\"int\":\"0\", default:\"null\"] "+newline+
                "var(type,name) ::= \"<type> <name> = <typeInit.(type)>;\""+newline
                ;
        Misc.writeFile(tmpdir, "test.stg", templates);
        STGroup group = STGroup.load(tmpdir+"/"+"test.stg");
        ST st = group.getInstanceOf("var");
        st.add("type", "UserRecord");
        st.add("name", "x");
        String expecting = "UserRecord x = null;";
        String result = st.render();
        assertEquals(expecting, result);
    }

    @Test public void testDictEmptyDefaultValue() throws Exception {
        String templates =
                "group test;" +newline+
                "typeInit ::= [\"int\":\"0\", default:] "+newline+
                "var(type,name) ::= \"<type> <name> = <typeInit.(type)>;\""+newline
                ;
        Misc.writeFile(tmpdir, "test.stg", templates);
        STGroup group = STGroup.load(tmpdir+"/"+"test.stg");
        ST st = group.getInstanceOf("var");
        st.add("type", "UserRecord");
        st.add("name", "x");
        String expecting = "UserRecord x = ;";
        String result = st.render();
        assertEquals(expecting, result);
    }

    @Test public void testDictDefaultValueIsKey() throws Exception {
        String templates =
                "group test;" +newline+
                "typeInit ::= [\"int\":\"0\", default:key] "+newline+
                "var(type,name) ::= \"<type> <name> = <typeInit.(type)>;\""+newline
                ;
        Misc.writeFile(tmpdir, "test.stg", templates);
        STGroup group = STGroup.load(tmpdir+"/"+"test.stg");
        ST st = group.getInstanceOf("var");
        st.add("type", "UserRecord");
        st.add("name", "x");
        String expecting = "UserRecord x = UserRecord;";
        String result = st.render();
        assertEquals(expecting, result);
    }

    /**
     * Test that a map can have only the default entry.
     * <p>
     * Bug ref: JIRA bug ST-15 (Fixed)
     */
    @Test public void testDictDefaultStringAsKey() throws Exception {
        String templates =
                "group test;" +newline+
                "typeInit ::= [\"default\":\"foo\"] "+newline+
                "var(type,name) ::= \"<type> <name> = <typeInit.(type)>;\""+newline
                ;
        Misc.writeFile(tmpdir, "test.stg", templates);
        STGroup group = STGroup.load(tmpdir+"/"+"test.stg");
        ST st = group.getInstanceOf("var");
        st.add("type", "default");
        st.add("name", "x");
        String expecting = "default x = foo;";
        String result = st.render();
        assertEquals(expecting, result);
    }
    
    /**
     * Test that a map can return a <b>string</b> with the word: default.
     * <p>
     * Bug ref: JIRA bug ST-15 (Fixed)
     */
    @Test public void testDictDefaultIsDefaultString() throws Exception {
        String templates =
                "group test;" +newline+
                "map ::= [default: \"default\"] "+newline+
                "t() ::= << <map.(\"1\")> >>"+newline                
                ;
        Misc.writeFile(tmpdir, "test.stg", templates);
        STGroup group = STGroup.load(tmpdir+"/"+"test.stg");
        ST st = group.getInstanceOf("t");
        String expecting = "default";
        String result = st.render();        
        assertEquals(expecting, result);
    }    

    @Test public void testDictViaEnclosingTemplates() throws Exception {
        String templates =
                "group test;" +newline+
                "typeInit ::= [\"int\":\"0\", \"float\":\"0.0\"] "+newline+
// TODO                "intermediate(type,name) ::= \"<var(...)>\""+newline+
                "intermediate(type,name) ::= \"<var()>\""+newline+
                "var(type,name) ::= \"<type> <name> = <typeInit.(type)>;\""+newline
                ;
        Misc.writeFile(tmpdir, "test.stg", templates);
        STGroup group = STGroup.load(tmpdir+"/"+"test.stg");
        ST st = group.getInstanceOf("intermediate");
        st.add("type", "int");
        st.add("name", "x");
        String expecting = "int x = 0;";
        String result = st.render();
        assertEquals(expecting, result);
    }

    @Test public void testDictViaEnclosingTemplates2() throws Exception {
        String templates =
                "group test;" +newline+
                "typeInit ::= [\"int\":\"0\", \"float\":\"0.0\"] "+newline+
                "intermediate(stuff) ::= \"<stuff>\""+newline+
                "var(type,name) ::= \"<type> <name> = <typeInit.(type)>;\""+newline
                ;
        Misc.writeFile(tmpdir, "test.stg", templates);
        STGroup group = STGroup.load(tmpdir+"/"+"test.stg");
        ST interm = group.getInstanceOf("intermediate");
        ST var = group.getInstanceOf("var");
        var.add("type", "int");
        var.add("name", "x");
        interm.add("stuff", var);
        String expecting = "int x = 0;";
        String result = interm.render();
        assertEquals(expecting, result);
    }    
}

File: objc/main/test/output/TestDictionaries.m
#import "TestDictionaries.h"

@implementation TestDictionaries

- (void) testDict {
  NSString * templates = [[@"typeInit ::= [\"int\":\"0\", \"float\":\"0.0\"] " stringByAppendingString:newline] stringByAppendingString:@"var(type,name) ::= \"<type> <name> = <typeInit.(type)>;\""] + newline;
  [self writeFile:tmpdir param1:@"test.stg" param2:templates];
  STGroup * group = [[[STGroupFile alloc] init:[[tmpdir stringByAppendingString:@"/"] stringByAppendingString:@"test.stg"]] autorelease];
  ST * st = [group getInstanceOf:@"var"];
  [st add:@"type" param1:@"int"];
  [st add:@"name" param1:@"x"];
  NSString * expecting = @"int x = 0;";
  NSString * result = [st render];
  [self assertEquals:expecting param1:result];
}

- (void) testDictValuesAreTemplates {
  NSString * templates = [[@"typeInit ::= [\"int\":{0<w>}, \"float\":{0.0<w>}] " stringByAppendingString:newline] stringByAppendingString:@"var(type,w,name) ::= \"<type> <name> = <typeInit.(type)>;\""] + newline;
  [self writeFile:tmpdir param1:@"test.stg" param2:templates];
  STGroup * group = [[[STGroupFile alloc] init:[[tmpdir stringByAppendingString:@"/"] stringByAppendingString:@"test.stg"]] autorelease];
  ST * st = [group getInstanceOf:@"var"];
  [st.impl dump];
  [st add:@"w" param1:@"L"];
  [st add:@"type" param1:@"int"];
  [st add:@"name" param1:@"x"];
  NSString * expecting = @"int x = 0L;";
  NSString * result = [st render];
  [self assertEquals:expecting param1:result];
}

- (void) testDictKeyLookupViaTemplate {
  NSString * templates = [[@"typeInit ::= [\"int\":{0<w>}, \"float\":{0.0<w>}] " stringByAppendingString:newline] stringByAppendingString:@"var(type,w,name) ::= \"<type> <name> = <typeInit.(type)>;\""] + newline;
  [self writeFile:tmpdir param1:@"test.stg" param2:templates];
  STGroup * group = [[[STGroupFile alloc] init:[[tmpdir stringByAppendingString:@"/"] stringByAppendingString:@"test.stg"]] autorelease];
  ST * st = [group getInstanceOf:@"var"];
  [st add:@"w" param1:@"L"];
  [st add:@"type" param1:[[[ST alloc] init:@"int"] autorelease]];
  [st add:@"name" param1:@"x"];
  NSString * expecting = @"int x = 0L;";
  NSString * result = [st render];
  [self assertEquals:expecting param1:result];
}

- (void) testDictKeyLookupAsNonToStringableObject {
  NSString * templates = [@"foo(m,k) ::= \"<m.(k)>\"" stringByAppendingString:newline];
  [self writeFile:tmpdir param1:@"test.stg" param2:templates];
  STGroup * group = [[[STGroupFile alloc] init:[[tmpdir stringByAppendingString:@"/"] stringByAppendingString:@"test.stg"]] autorelease];
  ST * st = [group getInstanceOf:@"foo"];
  NSMutableDictionary * m = [[[NSMutableDictionary alloc] init] autorelease];
  [m setObject:[[[HashableUser alloc] init:99 param1:@"parrt"] autorelease] param1:@"first"];
  [m setObject:[[[HashableUser alloc] init:172036 param1:@"tombu"] autorelease] param1:@"second"];
  [m setObject:[[[HashableUser alloc] init:391 param1:@"sriram"] autorelease] param1:@"third"];
  [st add:@"m" param1:m];
  [st add:@"k" param1:[[[HashableUser alloc] init:172036 param1:@"tombu"] autorelease]];
  NSString * expecting = @"second";
  NSString * result = [st render];
  [self assertEquals:expecting param1:result];
}

- (void) testDictMissingDefaultValueIsEmpty {
  NSString * templates = [[@"typeInit ::= [\"int\":\"0\", \"float\":\"0.0\"] " stringByAppendingString:newline] stringByAppendingString:@"var(type,w,name) ::= \"<type> <name> = <typeInit.(type)>;\""] + newline;
  [self writeFile:tmpdir param1:@"test.stg" param2:templates];
  STGroup * group = [[[STGroupFile alloc] init:[[tmpdir stringByAppendingString:@"/"] stringByAppendingString:@"test.stg"]] autorelease];
  ST * st = [group getInstanceOf:@"var"];
  [st add:@"w" param1:@"L"];
  [st add:@"type" param1:@"double"];
  [st add:@"name" param1:@"x"];
  NSString * expecting = @"double x = ;";
  NSString * result = [st render];
  [self assertEquals:expecting param1:result];
}

- (void) testDictMissingDefaultValueIsEmptyForNullKey {
  NSString * templates = [[@"typeInit ::= [\"int\":\"0\", \"float\":\"0.0\"] " stringByAppendingString:newline] stringByAppendingString:@"var(type,w,name) ::= \"<type> <name> = <typeInit.(type)>;\""] + newline;
  [self writeFile:tmpdir param1:@"test.stg" param2:templates];
  STGroup * group = [[[STGroupFile alloc] init:[[tmpdir stringByAppendingString:@"/"] stringByAppendingString:@"test.stg"]] autorelease];
  ST * st = [group getInstanceOf:@"var"];
  [st add:@"w" param1:@"L"];
  [st add:@"type" param1:nil];
  [st add:@"name" param1:@"x"];
  NSString * expecting = @" x = ;";
  NSString * result = [st render];
  [self assertEquals:expecting param1:result];
}

- (void) testDictHiddenByFormalArg {
  NSString * templates = [[@"typeInit ::= [\"int\":\"0\", \"float\":\"0.0\"] " stringByAppendingString:newline] stringByAppendingString:@"var(typeInit,type,name) ::= \"<type> <name> = <typeInit.(type)>;\""] + newline;
  [self writeFile:tmpdir param1:@"test.stg" param2:templates];
  STGroup * group = [[[STGroupFile alloc] init:[[tmpdir stringByAppendingString:@"/"] stringByAppendingString:@"test.stg"]] autorelease];
  ST * st = [group getInstanceOf:@"var"];
  [st add:@"type" param1:@"int"];
  [st add:@"name" param1:@"x"];
  NSString * expecting = @"int x = ;";
  NSString * result = [st render];
  [self assertEquals:expecting param1:result];
}

- (void) testDictEmptyValueAndAngleBracketStrings {
  NSString * templates = [[@"typeInit ::= [\"int\":\"0\", \"float\":, \"double\":<<0.0L>>] " stringByAppendingString:newline] stringByAppendingString:@"var(type,name) ::= \"<type> <name> = <typeInit.(type)>;\""] + newline;
  [self writeFile:tmpdir param1:@"test.stg" param2:templates];
  STGroup * group = [[[STGroupFile alloc] init:[[tmpdir stringByAppendingString:@"/"] stringByAppendingString:@"test.stg"]] autorelease];
  ST * st = [group getInstanceOf:@"var"];
  [st add:@"type" param1:@"float"];
  [st add:@"name" param1:@"x"];
  NSString * expecting = @"float x = ;";
  NSString * result = [st render];
  [self assertEquals:expecting param1:result];
}

- (void) testDictDefaultValue {
  NSString * templates = [[@"typeInit ::= [\"int\":\"0\", default:\"null\"] " stringByAppendingString:newline] stringByAppendingString:@"var(type,name) ::= \"<type> <name> = <typeInit.(type)>;\""] + newline;
  [self writeFile:tmpdir param1:@"test.stg" param2:templates];
  STGroup * group = [[[STGroupFile alloc] init:[[tmpdir stringByAppendingString:@"/"] stringByAppendingString:@"test.stg"]] autorelease];
  ST * st = [group getInstanceOf:@"var"];
  [st add:@"type" param1:@"UserRecord"];
  [st add:@"name" param1:@"x"];
  NSString * expecting = @"UserRecord x = null;";
  NSString * result = [st render];
  [self assertEquals:expecting param1:result];
}

- (void) testDictNullKeyGetsDefaultValue {
  NSString * templates = [[@"typeInit ::= [\"int\":\"0\", default:\"null\"] " stringByAppendingString:newline] stringByAppendingString:@"var(type,name) ::= \"<type> <name> = <typeInit.(type)>;\""] + newline;
  [self writeFile:tmpdir param1:@"test.stg" param2:templates];
  STGroup * group = [[[STGroupFile alloc] init:[[tmpdir stringByAppendingString:@"/"] stringByAppendingString:@"test.stg"]] autorelease];
  ST * st = [group getInstanceOf:@"var"];
  [st add:@"name" param1:@"x"];
  NSString * expecting = @" x = null;";
  NSString * result = [st render];
  [self assertEquals:expecting param1:result];
}

- (void) testDictEmptyDefaultValue {
  NSString * templates = [[@"typeInit ::= [\"int\":\"0\", default:] " stringByAppendingString:newline] stringByAppendingString:@"var(type,name) ::= \"<type> <name> = <typeInit.(type)>;\""] + newline;
  [self writeFile:tmpdir param1:@"test.stg" param2:templates];
  ErrorBuffer * errors = [[[ErrorBuffer alloc] init] autorelease];
  STGroupFile * group = [[[STGroupFile alloc] init:[[tmpdir stringByAppendingString:@"/"] stringByAppendingString:@"test.stg"]] autorelease];
  [group setListener:errors];
  [group load];
  NSString * expected = @"[test.stg 1:33: missing value for key at ']']";
  NSString * result = [errors.errors description];
  [self assertEquals:expected param1:result];
}

- (void) testDictDefaultValueIsKey {
  NSString * templates = [[@"typeInit ::= [\"int\":\"0\", default:key] " stringByAppendingString:newline] stringByAppendingString:@"var(type,name) ::= \"<type> <name> = <typeInit.(type)>;\""] + newline;
  [self writeFile:tmpdir param1:@"test.stg" param2:templates];
  STGroup * group = [[[STGroupFile alloc] init:[[tmpdir stringByAppendingString:@"/"] stringByAppendingString:@"test.stg"]] autorelease];
  ST * st = [group getInstanceOf:@"var"];
  [st add:@"type" param1:@"UserRecord"];
  [st add:@"name" param1:@"x"];
  NSString * expecting = @"UserRecord x = UserRecord;";
  NSString * result = [st render];
  [self assertEquals:expecting param1:result];
}


/**
 * Test that a map can have only the default entry.
 */
- (void) testDictDefaultStringAsKey {
  NSString * templates = [[@"typeInit ::= [\"default\":\"foo\"] " stringByAppendingString:newline] stringByAppendingString:@"var(type,name) ::= \"<type> <name> = <typeInit.(type)>;\""] + newline;
  [self writeFile:tmpdir param1:@"test.stg" param2:templates];
  STGroup * group = [[[STGroupFile alloc] init:[[tmpdir stringByAppendingString:@"/"] stringByAppendingString:@"test.stg"]] autorelease];
  ST * st = [group getInstanceOf:@"var"];
  [st add:@"type" param1:@"default"];
  [st add:@"name" param1:@"x"];
  NSString * expecting = @"default x = foo;";
  NSString * result = [st render];
  [self assertEquals:expecting param1:result];
}


/**
 * Test that a map can return a <b>string</b> with the word: default.
 */
- (void) testDictDefaultIsDefaultString {
  NSString * templates = [[@"map ::= [default: \"default\"] " stringByAppendingString:newline] stringByAppendingString:@"t() ::= << <map.(\"1\")> >>"] + newline;
  [self writeFile:tmpdir param1:@"test.stg" param2:templates];
  STGroup * group = [[[STGroupFile alloc] init:[[tmpdir stringByAppendingString:@"/"] stringByAppendingString:@"test.stg"]] autorelease];
  ST * st = [group getInstanceOf:@"t"];
  NSString * expecting = @" default ";
  NSString * result = [st render];
  [self assertEquals:expecting param1:result];
}

- (void) testDictViaEnclosingTemplates {
  NSString * templates = [[[@"typeInit ::= [\"int\":\"0\", \"float\":\"0.0\"] " stringByAppendingString:newline] stringByAppendingString:@"intermediate(type,name) ::= \"<var(type,name)>\""] + newline stringByAppendingString:@"var(type,name) ::= \"<type> <name> = <typeInit.(type)>;\""] + newline;
  [self writeFile:tmpdir param1:@"test.stg" param2:templates];
  STGroup * group = [[[STGroupFile alloc] init:[[tmpdir stringByAppendingString:@"/"] stringByAppendingString:@"test.stg"]] autorelease];
  ST * st = [group getInstanceOf:@"intermediate"];
  [st add:@"type" param1:@"int"];
  [st add:@"name" param1:@"x"];
  NSString * expecting = @"int x = 0;";
  NSString * result = [st render];
  [self assertEquals:expecting param1:result];
}

- (void) testDictViaEnclosingTemplates2 {
  NSString * templates = [[[@"typeInit ::= [\"int\":\"0\", \"float\":\"0.0\"] " stringByAppendingString:newline] stringByAppendingString:@"intermediate(stuff) ::= \"<stuff>\""] + newline stringByAppendingString:@"var(type,name) ::= \"<type> <name> = <typeInit.(type)>;\""] + newline;
  [self writeFile:tmpdir param1:@"test.stg" param2:templates];
  STGroup * group = [[[STGroupFile alloc] init:[[tmpdir stringByAppendingString:@"/"] stringByAppendingString:@"test.stg"]] autorelease];
  ST * interm = [group getInstanceOf:@"intermediate"];
  ST * var = [group getInstanceOf:@"var"];
  [var add:@"type" param1:@"int"];
  [var add:@"name" param1:@"x"];
  [interm add:@"stuff" param1:var];
  NSString * expecting = @"int x = 0;";
  NSString * result = [interm render];
  [self assertEquals:expecting param1:result];
}

- (void) TestAccessDictionaryFromAnonymousTemplate {
  NSString * dir = tmpdir;
  NSString * g = [[[[@"a() ::= <<[<[\"foo\",\"a\"]:{x|<if(values.(x))><x><endif>}>]>>\n" stringByAppendingString:@"values ::= [\n"] stringByAppendingString:@"    \"a\":false,\n"] stringByAppendingString:@"    default:true\n"] stringByAppendingString:@"]\n"];
  [self writeFile:dir param1:@"g.stg" param2:g];
  STGroup * group = [[[STGroupFile alloc] init:[[tmpdir stringByAppendingString:@"/"] stringByAppendingString:@"g.stg"]] autorelease];
  ST * st = [group getInstanceOf:@"a"];
  NSString * expected = @"[foo]";
  NSString * result = [st render];
  [self assertEquals:expected param1:result];
}

- (void) TestAccessDictionaryFromAnonymousTemplateInRegion {
  NSString * dir = tmpdir;
  NSString * g = [[[[[[[@"a() ::= <<[<@r()>]>>\n" stringByAppendingString:@"@a.r() ::= <<\n"] stringByAppendingString:@"<[\"foo\",\"a\"]:{x|<if(values.(x))><x><endif>}>\n"] stringByAppendingString:@">>\n"] stringByAppendingString:@"values ::= [\n"] stringByAppendingString:@"    \"a\":false,\n"] stringByAppendingString:@"    default:true\n"] stringByAppendingString:@"]\n"];
  [self writeFile:dir param1:@"g.stg" param2:g];
  STGroup * group = [[[STGroupFile alloc] init:[[tmpdir stringByAppendingString:@"/"] stringByAppendingString:@"g.stg"]] autorelease];
  ST * st = [group getInstanceOf:@"a"];
  NSString * expected = @"[foo]";
  NSString * result = [st render];
  [self assertEquals:expected param1:result];
}

@end
